### ISSUE TITLE
CNTK 2.6.0-rc0.dev20180821 with .NET Core 2.1 throws StackOverflowException when Trainer.TrainMinibatch() is called

### ISSUE THREAD

**Issue by kkc0923** [Microsoft/CNTK#3368]
Here is reproducible code:

program.cs
```csharp
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using CNTK;

namespace CNTKExceptionSample
{
    class Program
    {
        static void Main(string[] args)
        {
            DeviceDescriptor.TrySetDefaultDevice(DeviceDescriptor.GPUDevice(0), true);
            DeviceDescriptor device = DeviceDescriptor.UseDefaultDevice();

            Function model = Function.Load("model.cntk", device);
            Variable inputVariable = model.Arguments.Single();
            Variable outputVariable = model.Output;
            Variable labelVariable = Variable.InputVariable(model.Output.Shape, DataType.Float, dynamicAxes: new List<Axis> { Axis.DefaultBatchAxis() });

            int inputDim = inputVariable.Shape.Dimensions.Aggregate((d1, d2) => d1 * d2);
            int outputDim = outputVariable.Shape.Dimensions.Aggregate((d1, d2) => d1 * d2);

            float[] inputBuffer = new float[inputDim];
            float[] labelBuffer = new float[outputDim];

            Value inputValue = Value.CreateBatch<float>(inputVariable.Shape, inputBuffer, device);
            Value labelValue = Value.CreateBatch<float>(labelVariable.Shape, labelBuffer, device);
            
            Function lossFunction = CNTKLib.BinaryCrossEntropy(outputVariable, labelVariable);
            Function errorFunction = CNTKLib.SquaredError(outputVariable, labelVariable);

            Trainer trainer = Trainer.CreateTrainer(model, lossFunction, errorFunction, new List<Learner> {
                    CNTKLib.AdamLearner(
                        new ParameterVector(model.Parameters().ToArray()),
                        new TrainingParameterScheduleDouble(0.001),
                        new TrainingParameterScheduleDouble(0.9))
                });

            Dictionary<Variable, Value> minibatchDictionary = new Dictionary<Variable, Value>
            {
                [inputVariable] = inputValue,
                [labelVariable] = labelValue
            };

            int maxMinibatchCount = 100;
            int currentMinibatchCount = 0;

            while (currentMinibatchCount < maxMinibatchCount)
            {
                trainer.TrainMinibatch(minibatchDictionary, false, device);

                double loss = trainer.PreviousMinibatchLossAverage();

                if (loss < 1e-5)
                {
                    break;
                }

                Console.WriteLine($"Current minibatch={currentMinibatchCount}, loss={loss}");

                currentMinibatchCount++;
            }

            model.Save("trained-model.cntk");
        }
    }
}
```
resnet_models.py
```python
# Copyright (c) Microsoft. All rights reserved.
#
# Licensed under the MIT license. See LICENSE.md file in the project root
# for full license information.
# ==============================================================================

import numpy as np
from cntk.initializer import he_normal, normal
from cntk.layers import AveragePooling, MaxPooling, BatchNormalization, Convolution, Dense
from cntk.ops import element_times, relu

#
# assembly components
#


def conv_bn(input, filter_size, num_filters, strides=(1, 1), init=he_normal(), bn_init_scale=1):
    c = Convolution(filter_size, num_filters, activation=None,
                    init=init, pad=True, strides=strides, bias=False)(input)
    r = BatchNormalization(map_rank=1, normalization_time_constant=4096,
                           use_cntk_engine=False, init_scale=bn_init_scale, disable_regularization=True)(c)
    return r


def conv_bn_relu(input, filter_size, num_filters, strides=(1, 1), init=he_normal()):
    r = conv_bn(input, filter_size, num_filters, strides, init, 1)
    return relu(r)

#
# ResNet components
#


def resnet_basic(input, num_filters):
    c1 = conv_bn_relu(input, (3, 3), num_filters)
    c2 = conv_bn(c1, (3, 3), num_filters, bn_init_scale=1)
    p = c2 + input
    return relu(p)


def resnet_basic_inc(input, num_filters, strides=(2, 2)):
    c1 = conv_bn_relu(input, (3, 3), num_filters, strides)
    c2 = conv_bn(c1, (3, 3), num_filters, bn_init_scale=1)
    s = conv_bn(input, (1, 1), num_filters, strides)  # Shortcut
    p = c2 + s
    return relu(p)


def resnet_basic_stack(input, num_stack_layers, num_filters):
    assert(num_stack_layers >= 0)
    l = input
    for _ in range(num_stack_layers):
        l = resnet_basic(l, num_filters)
    return l


def resnet_bottleneck(input, out_num_filters, inter_out_num_filters):
    c1 = conv_bn_relu(input, (1, 1), inter_out_num_filters)
    c2 = conv_bn_relu(c1, (3, 3), inter_out_num_filters)
    c3 = conv_bn(c2, (1, 1), out_num_filters, bn_init_scale=0)
    p = c3 + input
    return relu(p)


def resnet_bottleneck_inc(input, out_num_filters, inter_out_num_filters, stride1x1, stride3x3):
    c1 = conv_bn_relu(input, (1, 1), inter_out_num_filters, strides=stride1x1)
    c2 = conv_bn_relu(c1, (3, 3), inter_out_num_filters, strides=stride3x3)
    c3 = conv_bn(c2, (1, 1), out_num_filters, bn_init_scale=0)
    stride = np.multiply(stride1x1, stride3x3)
    s = conv_bn(input, (1, 1), out_num_filters, strides=stride)  # Shortcut
    p = c3 + s
    return relu(p)


def resnet_bottleneck_stack(input, num_stack_layers, out_num_filters, inter_out_num_filters):
    assert(num_stack_layers >= 0)
    l = input
    for _ in range(num_stack_layers):
        l = resnet_bottleneck(l, out_num_filters, inter_out_num_filters)
    return l

#
# Defines the residual network model for classifying images
#


def create_cifar10_model(input, num_stack_layers, num_classes):
    c_map = [16, 32, 64]

    conv = conv_bn_relu(input, (3, 3), c_map[0])
    r1 = resnet_basic_stack(conv, num_stack_layers, c_map[0])

    r2_1 = resnet_basic_inc(r1, c_map[1])
    r2_2 = resnet_basic_stack(r2_1, num_stack_layers-1, c_map[1])

    r3_1 = resnet_basic_inc(r2_2, c_map[2])
    r3_2 = resnet_basic_stack(r3_1, num_stack_layers-1, c_map[2])

    # Global average pooling and output
    pool = AveragePooling(filter_shape=(8, 8), name='final_avg_pooling')(r3_2)
    z = Dense(num_classes, init=normal(0.01))(pool)
    return z


def create_imagenet_model_basic(input, num_stack_layers, num_classes):
    c_map = [64, 128, 256, 512]

    conv = conv_bn_relu(input, (7, 7), c_map[0], strides=(2, 2))
    pool1 = MaxPooling((3, 3), strides=(2, 2), pad=True)(conv)
    r1 = resnet_basic_stack(pool1, num_stack_layers[0], c_map[0])

    r2_1 = resnet_basic_inc(r1, c_map[1])
    r2_2 = resnet_basic_stack(r2_1, num_stack_layers[1], c_map[1])

    r3_1 = resnet_basic_inc(r2_2, c_map[2])
    r3_2 = resnet_basic_stack(r3_1, num_stack_layers[2], c_map[2])

    r4_1 = resnet_basic_inc(r3_2, c_map[3])
    r4_2 = resnet_basic_stack(r4_1, num_stack_layers[3], c_map[3])

    # Global average pooling and output
    pool = AveragePooling(filter_shape=(7, 7), name='final_avg_pooling')(r4_2)
    z = Dense(num_classes, init=normal(0.01))(pool)
    return z


def create_imagenet_model_bottleneck(input, num_stack_layers, num_classes, stride1x1, stride3x3):
    c_map = [64, 128, 256, 512, 1024, 2048]

    # conv1 and max pooling
    conv1 = conv_bn_relu(input, (7, 7), c_map[0], strides=(2, 2))
    pool1 = MaxPooling((3, 3), strides=(2, 2), pad=True)(conv1)

    # conv2_x
    r2_1 = resnet_bottleneck_inc(pool1, c_map[2], c_map[0], (1, 1), (1, 1))
    r2_2 = resnet_bottleneck_stack(
        r2_1, num_stack_layers[0], c_map[2], c_map[0])

    # conv3_x
    r3_1 = resnet_bottleneck_inc(
        r2_2, c_map[3], c_map[1], stride1x1, stride3x3)
    r3_2 = resnet_bottleneck_stack(
        r3_1, num_stack_layers[1], c_map[3], c_map[1])

    # conv4_x
    r4_1 = resnet_bottleneck_inc(
        r3_2, c_map[4], c_map[2], stride1x1, stride3x3)
    r4_2 = resnet_bottleneck_stack(
        r4_1, num_stack_layers[2], c_map[4], c_map[2])

    # conv5_x
    r5_1 = resnet_bottleneck_inc(
        r4_2, c_map[5], c_map[3], stride1x1, stride3x3)
    r5_2 = resnet_bottleneck_stack(
        r5_1, num_stack_layers[3], c_map[5], c_map[3])

    # Global average pooling and output
    pool = AveragePooling(filter_shape=(7, 7), name='final_avg_pooling')(r5_2)
    z = Dense(num_classes, init=normal(0.01))(pool)
    return z


def create_imagenet_model_bottleneck_pool(input, num_stack_layers, stride1x1, stride3x3):
    c_map = [64, 128, 256, 512, 1024, 2048]

    # conv1 and max pooling
    conv1 = conv_bn_relu(input, (7, 7), c_map[0], strides=(2, 2))
    pool1 = MaxPooling((3, 3), strides=(2, 2), pad=True)(conv1)

    # conv2_x
    r2_1 = resnet_bottleneck_inc(pool1, c_map[2], c_map[0], (1, 1), (1, 1))
    r2_2 = resnet_bottleneck_stack(
        r2_1, num_stack_layers[0], c_map[2], c_map[0])

    # conv3_x
    r3_1 = resnet_bottleneck_inc(
        r2_2, c_map[3], c_map[1], stride1x1, stride3x3)
    r3_2 = resnet_bottleneck_stack(
        r3_1, num_stack_layers[1], c_map[3], c_map[1])

    # conv4_x
    r4_1 = resnet_bottleneck_inc(
        r3_2, c_map[4], c_map[2], stride1x1, stride3x3)
    r4_2 = resnet_bottleneck_stack(
        r4_1, num_stack_layers[2], c_map[4], c_map[2])

    # conv5_x
    r5_1 = resnet_bottleneck_inc(
        r4_2, c_map[5], c_map[3], stride1x1, stride3x3)
    r5_2 = resnet_bottleneck_stack(
        r5_1, num_stack_layers[3], c_map[5], c_map[3])

    # Global average pooling and output
    pool = AveragePooling(filter_shape=(7, 7), name='final_avg_pooling')(r5_2)

    return pool
```
program.py
```python
import cntk as C
from cntk.initializer import he_normal, normal
from cntk.layers import (AveragePooling, BatchNormalization, Convolution,
                         Dense, MaxPooling)

import resnet_models


def create_resnet152(input, output_dim):
    stride1x1 = (1, 1)
    stride3x3 = (2, 2)

    z = input / 255

    z = resnet_models.create_imagenet_model_bottleneck_pool(
        z, [2, 7, 35, 2], stride1x1, stride3x3)

    with C.layers.default_options(init=normal(0.01), activation=C.leaky_relu):
        z = Dense(4000)(z)
        z = C.layers.Dropout(dropout_rate=0.5)(z)

    z = Dense(output_dim, init=normal(0.01), name='final_dense')(z)
    z = C.sigmoid(z)

    return z


input_variable = C.input_variable((3, 224, 224), name='input')
output_dim = 6310

model = create_resnet152(input_variable, output_dim)

model.save("model.cntk")
```

Reproduction step:
1. generate model.cntk using python code. (you should have installed cntk-gpu 2.5.1 via pip)
`python program.py`
2. copy generated model (model.cntk) to .NET Core 2.1 Console project. (you should install CNTK.GPU CNTK 2.6.0-rc0.dev20180821 from nuget)
3. overwrite program.cs using above source.
3. Run C# project.
You can see stackoverflow exception. 

If you create .Net framework (4.6.1 or 4.7.1) Console project with (stable CNTK.GPU 2.5.1 or 2.6.0-rc0.dev20180821) , no exception thrown and the model will be successfully trained.

**Comment by kkc0923** [comment]
CNTK.CPUOnly 2.6.0 + .NET Core 2.2 still occurred stackoverflow exception, when even evaluate pretrained model. Same source code and model works with .NET Framework 4.7.1.

**Comment by kkc0923** [comment]
@ebarsoumMS Any updates?

**Comment by Tixxx** [comment]
@kkc0923  apologize for missing this thread earlier. I looked through your code, the py model is quiet large so it likely puts a lot of pressure on the stack. Since net core 2.0 apps are built into dlls and invoked through dotnet.exe, it will take up more space on the stack compared to netframework. I would recommend you putting that main() code in your cs file into a separate thread and give it a large stack size and build with release config.  The default stack size is 1 MB, in the sample code below, i give it 2MB.
here's the modified code(tested on my dev box, it worked for me):
`
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading;
using CNTK;

namespace CNTKExceptionSample
{
    class Program
    {
        static void Main(string[] args)
        {
            // Start a thread that calls a parameterized static method and give it 2MB of stack size
            
            Thread newThread = new Thread(Train, 2097152);
            newThread.Start();
        }

        public static void Train()
        {
            DeviceDescriptor.TrySetDefaultDevice(DeviceDescriptor.CPUDevice, true);
            DeviceDescriptor device = DeviceDescriptor.UseDefaultDevice();

            Function model = Function.Load("model.cntk", device);
            Variable inputVariable = model.Arguments.Single();
            Variable outputVariable = model.Output;
            Variable labelVariable = Variable.InputVariable(model.Output.Shape, DataType.Float, dynamicAxes: new List<Axis> { Axis.DefaultBatchAxis() });

            int inputDim = inputVariable.Shape.Dimensions.Aggregate((d1, d2) => d1 * d2);
            int outputDim = outputVariable.Shape.Dimensions.Aggregate((d1, d2) => d1 * d2);

            float[] inputBuffer = new float[inputDim];
            float[] labelBuffer = new float[outputDim];

            Value inputValue = Value.CreateBatch<float>(inputVariable.Shape, inputBuffer, device);
            Value labelValue = Value.CreateBatch<float>(labelVariable.Shape, labelBuffer, device);

            Function lossFunction = CNTKLib.BinaryCrossEntropy(outputVariable, labelVariable);
            Function errorFunction = CNTKLib.SquaredError(outputVariable, labelVariable);

            Trainer trainer = Trainer.CreateTrainer(model, lossFunction, errorFunction, new List<Learner> {
                    CNTKLib.AdamLearner(
                        new ParameterVector(model.Parameters().ToArray()),
                        new TrainingParameterScheduleDouble(0.001),
                        new TrainingParameterScheduleDouble(0.9))
                });

            Dictionary<Variable, Value> minibatchDictionary = new Dictionary<Variable, Value>();
            minibatchDictionary.Add(inputVariable, inputValue);
            minibatchDictionary.Add(labelVariable, labelValue);

            int maxMinibatchCount = 10;
            int currentMinibatchCount = 0;

            while (currentMinibatchCount < maxMinibatchCount)
            {
                trainer.TrainMinibatch(minibatchDictionary, false, device);

                double loss = trainer.PreviousMinibatchLossAverage();

                if (loss < 1e-5)
                {
                    break;
                }

                Console.WriteLine($"Current minibatch={currentMinibatchCount}, loss={loss}");

                currentMinibatchCount++;
            }

            model.Save("trained-model.cntk");

        }
    }
}`

**Comment by kkc0923** [comment]
@Tixxx Thanks for reply. It seems to be only solution for resolving stack overflow. But it is not work for .NET's built-in threading methods, like Parallel, Task, ThreadPool and so on. When I serve my CNTK's network model and web-based evaluation interface, manually creating/managing thread is very hard and should be avoided. For this reason, I still used .NET Framework 4.7 instead of .NET Core. I hope that .NET Core should have option for changing default stack size when compiling.
